src/geo/transform.js

```diff
@@ -239,12 +239,36 @@ class Transform {
     }
 
     /**
-     * Given a point on screen, return its lnglat
+     * Given a point on screen and viewport options, return its lnglat
      * @param {Point} p screen point
+     * @param {CameraOptions} v viewport
      * @returns {LngLat} lnglat location
      */
-    pointLocation(p) {
-        return this.coordinateLocation(this.pointCoordinate(p));
+    pointLocation(p, v) {
+        if (v == null) {
+            return this.coordinateLocation(this.pointCoordinate(p));
+        } else {
+            const originalViewport = {
+                center: this.center,
+                zoom: this.zoom,
+                bearing: this.bearing,
+                pitch: this.pitch
+            };
+
+            if ('center' in v) this.center = v.center;
+            if ('zoom' in v) this.zoom = v.zoom;
+            if ('bearing' in v) this.bearing = v.bearing;
+            if ('pitch' in v) this.pitch = v.pitch;
```

Comment:
Ideally, we would work together to find a way to add a viewport parameter to `_calcMatrices`, reduce the amount of input needed by `_calcMatricies` where possible, and refactor to eliminate any duplicate code. Does that sound doable based on your experience in this part of the codebase?